File: hosts/simple/options.cc
/*
  Copyright 2007 Google Inc.

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/

#include "options.h"

Options::Options() {
}

Options::~Options() {
}

size_t Options::GetCount() {
  return values_.size();
}

void Options::Add(const char *name, const Variant &value) {
  if (values_.find(name) == values_.end()) {
    values_[name] = value;
    FireChangedEvent(name);
  }
}

bool Options::Exists(const char *name) {
  return values_.find(name) != values_.end();
}
  
Variant Options::GetDefaultValue(const char *name) {
  OptionsMap::const_iterator it = defaults_.find(name);
  return it == defaults_.end() ? Variant() : it->second;
}

void Options::PutDefaultValue(const char *name, const Variant &value) {
  defaults_[name] = value;
}

Variant Options::GetValue(const char *name) {
  OptionsMap::const_iterator it = values_.find(name);
  return it == values_.end() ? GetDefaultValue(name) : it->second;
}

void Options::PutValue(const char *name, const Variant &value) {
  values_[name] = value;
  FireChangedEvent(name);
}

void Options::Remove(const char *name) {
  OptionsMap::iterator it = values_.find(name);
  if (it != values_.end()) {
    values_.erase(it);
    FireChangedEvent(name);
  }
}

void Options::RemoveAll() {
  while (!values_.empty()) {
    OptionsMap::iterator it = values_.begin();
    std::string name(it->first);
    values_.erase(it);
    FireChangedEvent(name.c_str());
  }
}

void Options::FireChangedEvent(const char *name) {
  DLOG("option changed: %s", name);
}

File: trunk/igoogle_gadget/options.js
/*
  Copyright 2008 Google Inc.

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/

//  Array of user pref (name, value) pairs
var g_user_prefs = null;

/// Gadget init code.
opbrowser.onGetProperty = OnGetProperty;
opbrowser.onCallback = OnCallback;

// Options view

function OnGetProperty(p) {
  if (p == kSetPrefs) {
    return "\"function\"";
  }
}

function OnCallback(f, pref) {
  if (f == kSetPrefs) {
    gadget.debug.trace("SetPref " + pref);
    if (null == g_user_prefs) {
      g_user_prefs = new Array();
    }
    var index = Number(pref[0]) * 2;
    g_user_prefs[index] = pref[1];
    g_user_prefs[index + 1] = pref[2];
  }
}

function OnOK() {
  gadget.debug.trace("OK pressed");

  if (null != g_user_prefs) {
    for (var i = 0; i < g_user_prefs.length; i += 2) {
      var name = g_user_prefs[i];
      if (name) {
        var value = g_user_prefs[i + 1];
        options.putValue(name, value);
      }
    }

    g_user_prefs = null;
    // Toggle kRefreshOption
    var refresh = options.getValue(kRefreshOption);
    if (!refresh || refresh == "0") {
      options.putValue(kRefreshOption, "1");
    } else {
      options.putValue(kRefreshOption, "0");
    }
  }
}

function OnCancel() {
  // Do nothing
  gadget.debug.trace("cancel pressed");
}

function GeneratePage() {
  var code = options.getValue(kOptionsHTML);

  var html = "<html>" + kHTMLMeta;
  if (!code) {
    html += "<body><center><h6>" + strings.GADGET_NOTLOADED + "</h6></center>";
  }
  else {
    var preset = options.getValue(kPresetOptions);

    html += "<script type=\"text/javascript\">"
      + "function _gel(a){return document.getElementById(a);}"
      + "function OnChange(e){"
        + "var obj=e.target;"
        + "if (obj.value!=null){"
          +"window.external.SetPrefs(obj.id.substring(2,obj.id.length),"
          +"obj.name.substring(2,obj.name.length),obj.value);}}"
      + "function OnLoad(){"
        + "var lang=_gel('lang_selector');"
        + "if (lang!=null) lang.style.display='none';"
        + "var num_fields=_gel('m_numfields').value;"
        + "for(var i=0;i<num_fields;i++){"
          + "var obj=_gel('m_'+i);if(obj==null){continue;}"
          + preset
          + "obj.onchange=OnChange;obj.onclick=OnChange;}}"
      + "</script>"
      + "<body bgcolor=\"" + options.getvalue(kBGColorOption) 
      + "\" marginheight=\"0\" marginwidth=\"0\" onload=\"OnLoad()\">"
      + code;
  }
  html += "</body></html>";

  return html;
}

function OnOptionChanged() {
  if (event.propertyName == kOptionsHTML || 
      event.propertyName == kPresetOptions) {
    opbrowser.innerText = GeneratePage();
  }
}

function OnOpen() {
  opbrowser.innerText = GeneratePage();
}
